src/util/readwritefile.cpp

```diff
@@ -5,46 +5,45 @@
 
 #include <util/readwritefile.h>
 
+#include <streams.h>
 #include <util/fs.h>
 
 #include <algorithm>
 #include <cstdio>
 #include <limits>
 #include <string>
 #include <utility>
+#include <vector>
 
-std::pair<bool,std::string> ReadBinaryFile(const fs::path &filename, size_t maxsize)
+template <typename T>
+std::optional<T> ReadBinaryFile(const fs::path& filename, size_t maxsize)
 {
-    FILE *f = fsbridge::fopen(filename, "rb");
-    if (f == nullptr)
-        return std::make_pair(false,"");
-    std::string retval;
-    char buffer[128];
-    do {
-        const size_t n = fread(buffer, 1, std::min(sizeof(buffer), maxsize - retval.size()), f);
-        // Check for reading errors so we don't return any data if we couldn't
-        // read the entire file (or up to maxsize)
-        if (ferror(f)) {
-            fclose(f);
-            return std::make_pair(false,"");
-        }
-        retval.append(buffer, buffer+n);
-    } while (!feof(f) && retval.size() < maxsize);
-    fclose(f);
-    return std::make_pair(true,retval);
+    std::FILE *f = fsbridge::fopen(filename, "rb");
+    if (f == nullptr) return {};
+    T output{};
+    size_t file_size = fs::file_size(filename);
+    output.resize(std::min(file_size, maxsize));
+    try {
+        AutoFile{f} >> Span{output};
```

Comment:
I liked more the previous version which called `fread(3)` here. It was simple stupid. This `>>` is now hard to follow, especially given that it depends on `T`. For `vector` it ends up calling `AutoFile::detail_fread()`. It does not check whether `ferror(3)` occurred.

Reply:
> > It does not check whether ferror(3) occurred.
> 
> It does.

Where? There is no `ferror(3)` call. From the man page: "The function fread() does not distinguish between end-of-file and error, and callers must use feof(3) and ferror(3) to determine which occurred."

> If the return value of `detail_fread` is not `output.size()`, `operator>>` will fail.

Yes, but it can be equal to the desired size under two conditions: eof, or error. The previous code distinguished between both.

Reply:
That "must" in the man page is pretty clear: "callers must use feof(3) and ferror(3) to determine which occurred". A `ferror(3)` check can't hurt and it is better to have an extra check that always returns "no error" than a missing check, failing to detect an IO error. The previous code was doing that - a dumb `fread(3)` followed by an unconditional `ferror(3)`.

Reply:
I would extend the check to:

```diff
 void AutoFile::read(Span<std::byte> dst)
 {
-    if (detail_fread(dst) != dst.size()) {
+    if (detail_fread(dst) != dst.size() || std::ferror(m_file)) {
         throw std::ios_base::failure(feof() ? "AutoFile::read: end of file" : "AutoFile::read: fread failed");
```

Reply:
> if there is a bug in the current code in master, it should be fixed.

Done in https://github.com/bitcoin/bitcoin/pull/29307

Reply:
Glad something useful came out of this PR :-)